broadinstitute/single_cell_portal_core
# Using pre-built Docker image in CI/CD (SCP-4493)

### BACKGROUND
Currently, around 7.5 minutes is spent building a Docker image during a CI run to use for that job.  Most of this is running `bundle install` to install all required Ruby gems, and then the remainder running `yarn install` to install all required JavaScript packages.  In local Dockerized development, it can take as much as 25-30 minutes to build/spin up a local instance, making using Docker locally extremely unpalatable.

### CHANGES
This update introduces a pre-built Docker image that has all gems and packages pre-installed.  Now, deployed environments simply pull this image and call `docker run` directly via our convenience scripts.  This mounts the `node_modules` folder from the image into the project directory, allowing for JS components to be rendered directly via `rake assets:precompile` in deployed environments, or emitted via `vite` in local development.

In comparing CI runs before/after this change, there was a savings of ~4.5 minutes in startup time.  In local development, the time required to spin up a local instance fell by an order of magnitude (~25m to 2.5m).  While this hasn't been tested yet for a deployment, as there is additional work do be done in SCP-4494 to wire this into our CD pipeline, the expectation is that it will reduce the amount of actual downtime for the portal from ~5m to ~2.5m.

NOTE: Before this PR is merged, SCP-4494 will need to be addressed so that our deployment automation does not break due to missing Docker images.

### MANUAL TESTING
1. Pull this branch and make sure you have Docker running
2. Run the following command to boot the portal in Dockerized mode (this will pull the pre-built image using the `development` tag:
```
bin/load_env_secrets.sh -p secret/kdux/scp/development/$(whoami)/scp_config.json \
                        -r secret/kdux/scp/development/$(whoami)/read_only_service_account.json \
                        -s secret/kdux/scp/development/$(whoami)/scp_service_account.json
```
3. Confirm that you see the startup process complete in roughly 2-3 minutes (you can watch for the following text in `STDOUT`) and that you can bring the portal up locally at [https://localhost/single_cell](https://localhost/single_cell):
```
*** Runit started as PID 256
ok: run: /etc/service/nginx-log-forwarder: (pid 264) 0s
Jun 28 17:13:56 localhost cron[263]: (CRON) INFO (pidfile fd = 3)
Jun 28 17:13:56 localhost cron[263]: (CRON) INFO (Running @reboot jobs)
[ N 2022-06-28 17:13:56.1629 271/T1 age/Wat/WatchdogMain.cpp:1373 ]: Starting Passenger watchdog...
[ N 2022-06-28 17:13:56.1882 274/T1 age/Cor/CoreMain.cpp:1340 ]: Starting Passenger core...
[ N 2022-06-28 17:13:56.1909 274/T1 age/Cor/CoreMain.cpp:256 ]: Passenger core running in multi-application mode.
[ N 2022-06-28 17:13:56.2100 274/T1 age/Cor/CoreMain.cpp:1015 ]: Passenger core online, PID 274
[ N 2022-06-28 17:13:59.3326 274/T5 age/Cor/SecurityUpdateChecker.h:519 ]: Security update check: no update found (next check in 24 hours)
```
